Cache tools

mpranj opened the pull request:
Add tools for `kdb cache {enable,disable,clear}`. (as discussed in #2610)

## Basics

Check relevant points but **please do not remove entries**.
Do not describe the purpose of this PR in the PR description but:

- [x] Short descriptions should be in the release notes (added as entry in
      `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add something to the the release notes.**
- [ ] Longer descriptions should be in documentation or in design decisions.
- [x] Describe details of how you changed the code in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, should be in the commit messages.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests
- [x] I ran all tests locally and everything went fine
- [x] affected documentation is fixed
- [ ] I added code comments, logging, and assertions (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [x] meta data is updated (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))

## Review

Reviewers will usually check the following:

- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)
- [ ] Documentation is concise and good to read
- [ ] Examples are well chosen and understandable

## Labels

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the build server is happy** and also you
  say that everything is ready to be merged.

mpranj commented:
@markus2330 done.

markus2330 commented on src/libs/elektra/mount.c:
```diff
@@ -358,7 +358,7 @@ KeySet * elektraDefaultGlobalConfig (void)
 	// TODO: this is a poor way of detecting whether cache is compiled, but simply
 	// matching against cache might fail because of other plugins (e.g. "cachefilter")
-	if (strstr (ELEKTRA_PLUGINS, ";cache;") != NULL)
+	if (strstr (ELEKTRA_PLUGINS, ";cache;") != NULL && ksLookupByName (keys, "system/elektra/cache/enabled", 0))
```
I would remove the strstr part as it might be wrong (external compilation of cache plugin or similar). But this can be done in another PR.

markus2330 commented on src/plugins/cache/cache.c:
```diff
@@ -267,16 +289,29 @@ int elektraCacheGet (Plugin * handle ELEKTRA_UNUSED, KeySet * returned, Key * pa
+		ELEKTRA_LOG_DEBUG ("CLEAR CACHES path: %s", cacheFileName);
+
+		keySetString (cacheFile, cacheFileName);
+		nftw (cacheFileName, unlinkCacheFiles, MAX_FD_USED, FTW_DEPTH);
```
We should disable the plugin if nftw is not present see tests/cframework/CMakeLists.txt

markus2330 commented on src/tools/kdb/cache.cpp:
```diff
@@ -0,0 +1,79 @@
+	}
+	else
+	{
+		throw invalid_argument ("1 argument required");
```
error message could be better: write which options are available.

mpranj replied:
Throwing this exception already prints the synopsis, including the available options.

markus2330 replied:
Ok, then writing that cmd is not a valid subcommand should be enough.

markus2330 commented on src/tools/kdb/cache.cpp:
```diff
@@ -0,0 +1,79 @@
+		plugin->get (ks, errorKey);
+
+		printWarnings (cerr, errorKey, cl.verbose, cl.debug);
+		printError (cerr, errorKey, cl.verbose, cl.debug);
```
why print twice? (Is already printed in 72+73?)

mpranj replied:
It prints the warnings from a different KDB handle / parentKey.

markus2330 commented on src/tools/kdb/cache.hpp:
```diff
@@ -0,0 +1,44 @@
+	virtual int execute (Cmdline const & cmdline) override;
+};
+
+#endif
```
man pages in doc/help is missing.

markus2330 commented on src/plugins/cache/cache.c:
```diff
@@ -267,16 +289,29 @@ int elektraCacheGet (Plugin * handle ELEKTRA_UNUSED, KeySet * returned, Key * pa
 		ch->cacheStorage->global = elektraPluginGetGlobalKeySet (handle);
 	}
 
+	if (!elektraStrCmp (keyString (keyGetMeta (parentKey, "cacheClear")), "YES"))
```
cacheClear is not according our conventions, it should be cache/clear.

Furthermore it should be documented in doc/METADATA.ini

markus2330 commented on src/tools/kdb/cache.cpp:
```diff
@@ -0,0 +1,79 @@
+
+		KeySet ks;
+		Key errorKey ("system/elektra/cache/clear", KEY_END);
+		errorKey.setMeta ("cacheClear", "YES");
```
Please use the normalized version "1" for booleans.

markus2330 commented on src/plugins/cache/CMakeLists.txt:
```diff
@@ -14,6 +15,13 @@ if (NOT_INCLUDED)
+
+if (NOT HAVE_NFTW)
+	message ("nftw (ftw.h) not found, excluding the cache plugin")
+	return ()
```
Please exclude the plugin properly (with remove_plugin), otherwise it is not listed in the excluded plugins.

mpranj replied:
After adding the check with `safe_check_symbol_exists (nftw "ftw.h" HAVE_NFTW)`, it does not find nftw any more (on debian 9, where it works fine without the check).

mpranj commented:
I've implemented the requested changes. Unfortunately it seems that even issuing only a warning for the missing cache plugin already breaks [some tests](https://travis-ci.org/ElektraInitiative/libelektra/jobs/563909296). I'm not aware of another way to add the warning without breaking unit tests. Maybe we can completely silence it?

markus2330 commented:
Yes, silence it but only if the shared lib cannot be loaded and the cache was not enabled by your tool. Other problems with the cache should still lead to warnings.

I also created the issue #2834 because the text of the warning is wrong.

markus2330 commented:
Can you rebase this? I would like to have this PR for the next release.

mpranj commented:
> Can you rebase this? I would like to have this PR for the next release. (Ideally with the functionality as discussed.)

Sure, I'll work on it to get it done today with the discussed functionality!

markus2330 commented:
Btw. if you rebase, we use a new cmake-format (0.5.4 `pip install --upgrade cmake-format` should be enough to upgrade).

mpranj commented:
I already rebased, thank you for the hint.

I implemented the discussed changes. To recap the three new scenarios:

1. Default mode: try to mount cache, but fail silently if dlopen failed
2. Enabled: issue warning to the user
3. Disabled: plugin is not mounted anyway

I'm basically just waiting for the buildserver response now.

markus2330 commented:
Good to hear that!

There is a trivial problem reported from the build server:

```
diff --git a/src/plugins/cache/CMakeLists.txt b/src/plugins/cache/CMakeLists.txt
index 8653a7bac..57f14680f 100644
--- a/src/plugins/cache/CMakeLists.txt
+++ b/src/plugins/cache/CMakeLists.txt
@@ -33,4 +33,4 @@ if (CACHE_DEPENDENCIES_OK AND BUILD_SHARED)
 	add_dependencies (elektra-cache elektra-resolver_fm_hpu_b elektra-mmapstorage)
 endif ()
 
-unset(CACHE_DEPENDENCIES_OK)
+unset (CACHE_DEPENDENCIES_OK)
```

Did you use cmake-format 0.5.4 to format this?

markus2330 commented on doc/help/kdb-cache.md:
```diff
@@ -0,0 +1,45 @@
+
+## LIMITATIONS
+
+Caches are stored on a per-user basis, therefor the `clear`
```
```suggestion
Caches are stored on a per-user basis, therefore the `clear`
```

markus2330 commented on doc/help/kdb-cache.md:
```diff
@@ -0,0 +1,45 @@
+
+## SYNOPSIS
+
+`kdb cache {enable,disable,clear}`
```
What about a possibility to go back to the default?

mpranj replied:
Good idea, implemented!

mpranj commented:
> Did you use cmake-format 0.5.4 to format this?

I updated cmake-format, but it seems that I broke the formatting even more.

markus2330 commented:
I created a new branch  #2850 with one additional commit where I did reformat-cmake. Let us see if this helps.

mpranj commented:
There was an error in my installation... should be fixed now.

mpranj commented:
@markus2330 I put the changes together here. Somehow the updated kdb cache man-page is not regenerated even though the text changed. I don't really understand how it's decided when the man-pages are rebuilt. On my system it seems to happen randomly.

markus2330 commented:
Ok, thank you! Let us see if it builds now.
